micro-rdk/src/common/ota.rs

```diff
@@ -352,53 +363,97 @@ impl<S: OtaMetadataStorage> OtaService<S> {
+                ));
+            }
+            let mut sender = sender.unwrap();
+            let inner_conn = inner_conn.unwrap();
+
+            if retry {
+                log::debug!(
+                    "retry establishing http connection to {} in {} seconds",
+                    &uri,
+                    CONN_RETRY_SECS
+                );
+                Timer::after(Duration::from_secs(CONN_RETRY_SECS)).await;
+                continue;
             }
-        }));
-
-        log::info!("ota connected, beginning download");
-        let request = Request::builder()
-            .method("GET")
-            .uri(uri)
-            .body(Empty::<Bytes>::new())
-            .map_err(|e| OtaError::Other(e.to_string()))?;
-        let response = sender
-            .send_request(request)
-            .await
-            .map_err(|e| OtaError::Other(e.to_string()))?;
 
-        // TODO(RSDK-9346): handle other success codes
-        if response.status() != 200 {
-            return Err(OtaError::Other(
-                format!("Bad Request - Status:{}", response.status()).to_string(),
-            ));
+            conn = Some(Box::new(self.exec.spawn(async move {
+                if let Err(err) = inner_conn.await {
+                    log::error!("connection failed: {:?}", err);
+                }
+            })));
+
+            log::info!("ota connected, beginning download");
+            let request = Request::builder()
+                .method("GET")
+                .uri(uri)
+                .body(Empty::<Bytes>::new())
+                .map_err(|e| OtaError::Other(e.to_string()))?;
+            response = sender
+                .send_request(request)
+                .await
+                .map_err(|e| OtaError::Other(e.to_string()))?;
+
+            let status = response.status();
+            match (status.is_success(), status.is_redirection()) {
+                (true, false) => break,
+                (false, true) => {
+                    log::debug!("target url has been redirected...");
+                    let headers = response.headers();
+                    if !headers.contains_key(hyper::header::LOCATION) {
+                        log::error!("`location` not found in redirection response header");
+                        return Err(OtaError::Other(format!(
+                            "invalid redirection response header: {:?}",
+                            headers
+                        )));
+                    }
+                    uri = self.parse_uri(headers[hyper::header::LOCATION].to_str().map_err(
+                        |e| {
+                            OtaError::Other(format!(
+                                "invalid redirection `location` in header: {} - {:?}",
+                                e, headers
+                            ))
+                        },
+                    )?)?;
+                    drop(conn.take());
```

Comment:
- [question] Do you need the explicit `drop`? Wouldn't `let _  = conn.take()` do the trick?
- [minor] Log the redirected location at info here.

Reply:
It *should* but I know there are some cases where `let _ = {}` can get optimized out by the compiler. This was me being overly cautious if anything. can switch to that

Reply:
It's fine with `drop`.